### PR TITLE
[KYUUBI #5314] Support JDK 21

### LINKED ISSUE #5314: [TASK][EASY] Support JDK 21
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Skill requirements

- familiar with GitHub Action
- basic knowledge of Apache Maven, including profiles

### What would you like to be improved?

Java 21 and the JDK 21 release under a general availability (GA) status occurred Sep 19, 2023, as the newest major update to the Java programming language.

More details on Java 21 (JDK 21) GA release via the [JDK-dev mailing list](https://mail.openjdk.org/pipermail/jdk-dev/2023-September/008267.html). More details on all of the major improvements with Java 21 can be found via [OpenJDK.org](https://openjdk.org/projects/jdk/21/).

Kyuubi has supported Java 17 since 1.8, it would be nice if we could support Java 21.

### How should we improve?

Step 0: wait for https://github.com/apache/kyuubi/issues/6293
Step 1: test Kyuubi on JDK 21 and fix errors encountered
Step 2: add CI in GitHub Actions
Step 3(out of scope): leverage new features introduced in JDK 21, e.g. Virtual Thread

Spark 3.5 almost supports Java 21 (except for Arrow, see SPARK-45781), to pass Kyuubi's CI, we need to either 
- disable arrow-related tests; or
- replace arrow jars under Spark 3.5 runtime in CI workflow, just as we did for Hive https://github.com/apache/kyuubi/pull/5280

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by dupen01**
# :mag: Description
## Issue References 🔗
This pull request fixes #5314

## Describe Your Solution 🔧
To support Java 21, some dependency packages need to be upgraded.Here is the detailed explanation.
1. To build with Java 21, Scala binary version needs to be upgraded to 2.13.11 or above.
2. The versions of `spotless` and `googlejavaformat` need to be upgraded to higher versions. However, this will lose support for Java 8. In order to continue supporting Java 8, code style checks will be ignored in the Java 21 profile.
3. Since the Flink engine does not support Scala 2.13, the Scala version of the Flink module is fixed at 2.12.

Previous discussions were at https://github.com/apache/kyuubi/pull/6306.

---
## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

---

**Be nice. Be informative.**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1929,7 +1966,8 @@
             <id>scala-2.13</id>
             <properties>
                 <scala.binary.version>2.13</scala.binary.version>
-                <scala.version>2.13.8</scala.version>
+                <!-- Scala version above 2.13.11 support Java 21. -->
+                <scala.version>2.13.11</scala.version>
```
Comment (pan3793) [inline pom.xml:1970]: can you submit another PR to upgrade Scala 2.13 ahead?

File: externals/kyuubi-flink-sql-engine/pom.xml [hunk 2]
```diff
@@ -30,6 +30,12 @@
     <name>Kyuubi Project Engine Flink SQL</name>
     <url>https://kyuubi.apache.org/</url>
 
+    <!-- Keep Flink Engine with Scala-2.12. -->
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/pom.xml:33]: I know that, currently, the whole Kyuubi project can be compiled with Scala 2.13 except for the Flink engine, while considering that the official Kyuubi releases always use Scala 2.12, we don't need to make such changes.

Please revert Flink engine Scala version related changes.

In summary, what we need to do are:
1. upgrade Scala 2.13.11 in another PR ahead
2. just like what you do in https://github.com/apache/kyuubi/pull/6306, tune the root `pom.xml` and add profile `java-21`
3. enable Java 21 test on CI with Spark 4.0

Comment (dupen01) [reply externals/kyuubi-flink-sql-engine/pom.xml:33]: Thank you for your guidance, I will create a new PR to upgrade the Scala version to 2.13.11 soon

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master
